### DISCUSSION TITLE
Is disabling HSTS via a configuration parameter / environment variable possible?

### DISCUSSION THREAD

**Discussion by TekkertheChaot** [dani-garcia/vaultwarden#4634] [Q&A]
I have Vaultwarden set up in my local home network with my own CA for TLS for over a year now and am very happy with it! 
Only thing that almost drove me insane was the forced HSTS (Header `'strict-transport-security: max-age=31536000'`).
As I also run other containers on the same host _(who would have thought ;) )_ and not every container / web-interface supports TLS, I cannot use my other apps due to HSTS also trying to force HTTPS on other ports of the same host.

I've looked on how HSTS is managed in Rocket and found that the 'shield' Rust-module manages these settings and that most Rocket configs can be done via environment variables. 

Do you know of any possibility to disable HSTS through any configurable manner?

**Comment by BlackDex** [comment]
Vaultwarden doesn't configure this as far as i know. Now I'm not sure what happens if you configure Vaultwarden to use the certificates directly, but i can't remember it added that specific header. That would be strange also, since it could break other installations like you mentioned. 

Are you running a reverse proxy in front of it? If so, it probably is located there.

**Reply by stefan0xC** [reply]
Using `ROCKET_TLS={certs="data/cert.pem",key="data/key.pem"}` doesn't add that header on my local test setup:
```
HTTP/1.1 200 OK
content-type: text/html; charset=utf-8
cache-control: public, max-age=600
expires: Mon, 10 Jun 2024 06:16:02 GMT
server: Rocket
x-frame-options: SAMEORIGIN
permissions-policy: accelerometer=(), ambient-light-sensor=(), autoplay=(), battery=(), camera=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=(), execution-while-out-of-viewport=(), fullscreen=(), geolocation=(), gyroscope=(), keyboard-map=(), magnetometer=(), microphone=(), midi=(), payment=(), picture-in-picture=(), screen-wake-lock=(), sync-xhr=(), usb=(), web-share=(), xr-spatial-tracking=()
x-content-type-options: nosniff
referrer-policy: same-origin
x-xss-protection: 0
content-security-policy: default-src 'self'; base-uri 'self'; form-action 'self'; object-src 'self' blob:; script-src 'self' 'wasm-unsafe-eval'; style-src 'self' 'unsafe-inline'; child-src 'self' https://*.duosecurity.com https://*.duofederal.com; frame-src 'self' https://*.duosecurity.com https://*.duofederal.com; frame-ancestors 'self' chrome-extension://nngceckbapebfimnlniiiahkandclblb chrome-extension://jbkfoedolllekgbhcbcoahefnbanhhlh moz-extension://* ; img-src 'self' data: https://haveibeenpwned.com ; connect-src 'self' https://api.pwnedpasswords.com https://api.2fa.directory https://app.simplelogin.io/api/ https://app.addy.io/api/ https://api.fastmail.com/ https://api.forwardemail.net ;
content-length: 1236
date: Mon, 10 Jun 2024 06:06:02 GMT
```

**Reply by TekkertheChaot** [reply]
Thank you for the quick reply!
I have set up the certificates directly with the `'ROCKET_TLS={certs="XXX",key="XXX"}' `env variable without any reverse proxy.
AFAI understand this (with this config) ROCKET should be the only component involved in the communication. 🤔

In the meantime I was able to cross check if the HSTS header gets applied by from Vaultwarden.

1. Visit NAS dashboard via HTTP -> OK
2. Visit Vaultwarden web via HTTPS -> OK
3. Visit NAS dashboard via HTTP -> get redirected to HTTPS under same port

please correct me if im wrong in my methology

**Reply by TekkertheChaot** [reply]
> Using `ROCKET_TLS={certs="data/cert.pem",key="data/key.pem"}` doesn't add that header on my local test setup:

Hmmm that is weird... I will try it with another browser, maybe its a firefox thing or something...

**Reply by BlackDex** [reply]
Not sure how your setup is. But If the nas and Vaultwarden share an IP. Most browsers will force SSL if it ever was able to reach it on the default port. Some even have a force https option like Firefox. If that is enabled it will keep using https.

Maybe there is a reverse proxy in front of it that you do not know about? Handled by your nas?

**Reply by TekkertheChaot** [reply]
I just tried everything in ungoogled-chromium, unfortunately with the same behaviour but with a more extensive logging. 
![grafik](https://github.com/dani-garcia/vaultwarden/assets/17234581/8635be2a-f2ea-4ae9-9e94-efb4f2c32dc4)

**Reply by TekkertheChaot** [reply]
> Not sure how your setup is. But If the nas and Vaultwarden share an IP. Most browsers will force SSL if it ever was able to reach it on the default port. Some even have a force https option like Firefox. If that is enabled it will keep using https.
> 
> Maybe there is a reverse proxy in front of it that you do not know about? Handled by your nas?

THE TIMING! 
I will check it real quick

**Reply by TekkertheChaot** [reply]
Ok, just checked as you made a good point about how the browsers behave differently when reaching default ports. 
Both firefox and chromium behave the same on my end regarding this issue. I have also tested other apps not listening on HTTP/80 and HTTP/443. 
I mean Vaultwarden runs via HTTPS on a custom port (4744) on my end. Other services who do use HTTPS are not affected in my network, but they also didn't introduce HSTS. 🤔

**Reply by BlackDex** [reply]
> Using `ROCKET_TLS={certs="data/cert.pem",key="data/key.pem"}` doesn't add that header on my local test setup:
> 
> ```
> HTTP/1.1 200 OK
> content-type: text/html; charset=utf-8
> cache-control: public, max-age=600
> expires: Mon, 10 Jun 2024 06:16:02 GMT
> server: Rocket
> x-frame-options: SAMEORIGIN
> permissions-policy: accelerometer=(), ambient-light-sensor=(), autoplay=(), battery=(), camera=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=(), execution-while-out-of-viewport=(), fullscreen=(), geolocation=(), gyroscope=(), keyboard-map=(), magnetometer=(), microphone=(), midi=(), payment=(), picture-in-picture=(), screen-wake-lock=(), sync-xhr=(), usb=(), web-share=(), xr-spatial-tracking=()
> x-content-type-options: nosniff
> referrer-policy: same-origin
> x-xss-protection: 0
> content-security-policy: default-src 'self'; base-uri 'self'; form-action 'self'; object-src 'self' blob:; script-src 'self' 'wasm-unsafe-eval'; style-src 'self' 'unsafe-inline'; child-src 'self' https://*.duosecurity.com https://*.duofederal.com; frame-src 'self' https://*.duosecurity.com https://*.duofederal.com; frame-ancestors 'self' chrome-extension://nngceckbapebfimnlniiiahkandclblb chrome-extension://jbkfoedolllekgbhcbcoahefnbanhhlh moz-extension://* ; img-src 'self' data: https://haveibeenpwned.com ; connect-src 'self' https://api.pwnedpasswords.com https://api.2fa.directory https://app.simplelogin.io/api/ https://app.addy.io/api/ https://api.fastmail.com/ https://api.forwardemail.net ;
> content-length: 1236
> date: Mon, 10 Jun 2024 06:06:02 GMT
> ```

@stefan0xC could you try a non-debug build and see what happens?

**Reply by stefan0xC** [reply]
Ah, that's it. With `ROCKET_PROFILE=release` the header will be added.

**Reply by TekkertheChaot** [reply]
Ohhhh, that makes sense! 
Is it possible to just change `ROCKET_PROFILE` in my docker env or does vaultwarden need a rebuild?

**Reply by hiyan** [reply]
HSTS is enforced by the browser on a domain or host (not sure) level. Search for how to clear it in your browser.

Edge is - `edge://net-internals/#hsts`

**Reply by TekkertheChaot** [reply]
> HSTS is enforced by the browser on a domain or host (not sure) level. Search for how to clear it in your browser.
> 
> Edge is - `edge://net-internals/#hsts`

You are right, after HSTS is set I need to clear the website data / cache. 
My main browser is Firefox and it seems that there it isn't all that easy. Although there is a `about:config` flag called `network.stricttransportsecurity.preloadlist` it didn't seem to work on my end. And completely disabling it would just open a huge vulnerability.

**Reply by TekkertheChaot** [reply]
I've just tested the `ROCKET_PROFILE` variable in docker and set it to `debug` -> HSTS is not enforced anymore, yay!
But I am not sure what else is affected by the debug profile. 
Do you think this is a valid and "safe" workaround or would a HSTS-switch for Rocket:shield within the admin dashboard be more senseful?

**Reply by TekkertheChaot** [reply]
Bumping, just to be sure not to miss anything crucial: 
Do you think setting `ROCKET_PROFILE` to `debug` is a valid and "safe" workaround, or would a HSTS-switch for Rocket:shield within the admin dashboard be more sensible?

**Reply by TekkertheChaot** [reply]
@stefan0xC, sorry to ping you like this as there hasn't been any updates on this. Hope you are doing well!

Can you or somebody else verify if using the debug profile is "safe"? 
I'm running it in debug for over a month now and only noticed a slight bump in response times.

Alternatively, I would like this to be treated as a feature request (configuring HSTS via ENV or admin page) so it will not be overlooked. 😉
